### PR TITLE
feat: Video message

### PR DESCRIPTION
**Pull request by rarira**
## Specify project
Both client and server

## Description
Added uploading video feature 
Simulator Screen Recording - iPhone 13 - 2021-09-29 at 19 15 33

- Now we can add video file through both image picker and camera.
- Uploaded video file provide thumbnail at first and load async video files from azure cloud.
   Video thumbnail can't be provided on Web, because `expo-video-thumbnails` package doesn't work on Web.
   also on Web, `loadAsync` function of `expo-av` package doesn't work as expected so it loads video files as soon as loaded
- Now client side limit is 180seconds long and medium quality compressed (can change setting in `ImagePicker.ts`
- Server side limit is less than 100MB and can handle error code by express
- Added server side mime type check package `stream-mime-type` for providing azure uploaded blob with exact mime type, unless mobile OS (both android and ios) can't detect the downloaded file type which results in failing display. 

Lots of limitations which needs further updates
- `Web` related
   `Alert.alert()` is not working on Web. So all the errors on the web is not delivered to user at all. 
- `expo-av` picker bug
  https://github.com/expo/expo/issues/14129
  https://github.com/expo/expo/issues/13553 : have to stay 9.1.2 version until fixed
  https://github.com/expo/expo/issues/9374:  Very big size file crash the app, which error seems to be not propery handled in the package
- style issues of newly changed option UI.. especially on Web

## Related Issues
#479 

## Tests

I added the following tests:
Updates existing tests code and Added new test codes for newly added component: `VideoPlayer.tsx`

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/hackatalk/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] Run `yarn lint && yarn tsc`
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] I am willing to follow-up on review comments in a timely manner.

[wehack2021]-[cloud]

### CODE REVIEW

**Review by hyochan — CHANGES_REQUESTED** [review]
This is insanely great feature :high_brightness: 

Could you kindly check below warning in current `PR`? It happens when you enter in/out of [Message] page.




Also, please `pull` the code since I've made some updates for you on just rebasing and resolving conflicts.

File: server/src/apis/root.ts [hunk 1]
```diff
@@ -28,7 +30,8 @@ export function resolveBlobName(destFile: string, destDir: string): string {
 
 const router = Router();
 const storage = multer.memoryStorage();
-const upload = multer({storage});
+
+const upload = multer({storage, limits: {fileSize: 100000000}});
```
Comment (hyochan) [inline server/src/apis/root.ts:34]: I think the `fileSize` here can be written in `const.ts`.

File: client/src/components/uis/VideoPlayer.tsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            quality: 0.5,
+          });
+          setThumbnail(result);
+        } catch (e) {
+          setThumbnailError(getString('THUMBNAIL_ERROR'));
+        }
+      })();
+
+    return () => {
+      /* istanbul ignore else */
```
Comment (hyochan) [inline client/src/components/uis/VideoPlayer.tsx:73]: Could you kindly tell why you try to ignore the coverage here?

File: client/src/components/pages/Message.tsx [hunk 3]
```diff
@@ -203,56 +205,95 @@ const MessagesFragment: FC<MessageProp> = ({channelId, messages}) => {
...
+      }
 
-    if (image && !image.cancelled)
       try {
-        const resizedImage = await resizePhotoToMaxDimensionsAndCompressAsPNG({
-          uri: image.uri,
-          width: 1920,
-          height: 1920,
-        });
+        let response;
```
Comment (hyochan) [inline client/src/components/pages/Message.tsx:227]: Can we add `type` here?

File: client/src/components/pages/Message.tsx [hunk 4]
```diff
@@ -203,56 +205,95 @@ const MessagesFragment: FC<MessageProp> = ({channelId, messages}) => {
     });
   };
 
-  const onRequestImagePicker = async (type: string): Promise<void> => {
-    let image;
+  const onRequestMediaPicker = async (type: ImagePickerType): Promise<void> => {
+    let media;
```
Comment (hyochan) [inline client/src/components/pages/Message.tsx:209]: Similarly, can we add `type` here?

**Review by hyochan — CHANGES_REQUESTED** [review]

File: client/assets/langs/ko.json [hunk 5]
```diff
@@ -40,11 +41,13 @@
...
+  "LIMIT_FILE_SIZE": "100MB 이하의 파일만 업로드할 수 있습니다",
```
Comment (hyochan) [inline client/assets/langs/ko.json:44]: How about passing `{{videoFileSize}}` param after adding const variable in `const.ts`?

File: client/src/components/pages/Message.tsx [hunk 6]
```diff
@@ -203,56 +205,95 @@ const MessagesFragment: FC<MessageProp> = ({channelId, messages}) => {
...
+            media.uri,
+            'messages',
+            `_${channelId}_${new Date().toISOString()}`,
+          );
+        else {
+          const resizedImage = await resizePhotoToMaxDimensionsAndCompressAsPNG(
+            {
+              uri: media.uri,
+              width: 1920,
+              height: 1920,
```
Comment (hyochan) [inline client/src/components/pages/Message.tsx:240]: You may use const variables in `const.ts`.

File: client/src/components/uis/__tests__/MessageListItem.test.tsx [hunk 7]
```diff
@@ -53,25 +55,32 @@ const QueryWrapper: React.FC<QueryWrapperProps> = (props) => {
...
-  MockPayloadGenerator.generate(operation, {
-    Message(_, generateId) {
-      return {
-        ...SAMPLE_MESSAGE,
-        id: `test-message-${generateId()}`,
-      };
-    },
-  }),
-);
+// beforeEach(() => ();
```
Comment (hyochan) [inline client/src/components/uis/__tests__/MessageListItem.test.tsx:66]: I think we can remove this :)

**Comment by rarira** [comment]
> This is insanely great feature 🔆
> 
> Could you kindly check below warning in current `PR`? It happens when you enter in/out of [Message] page.
> 
> 
> 
> Also, please `pull` the code since I've made some updates for you on just rebasing and resolving conflicts.

https://github.com/expo/expo/issues/3115#issuecomment-509847661
I fixed this error by removing `unloadAsync` related lines as above comment suggested

File: client/src/components/uis/__tests__/MessageListItem.test.tsx (hunk 7 above)
Comment (rarira) [reply client/src/components/uis/__tests__/MessageListItem.test.tsx]: All above suggestion was applied. thans :)

**Review by hyochan — APPROVED** [review]
You've rocked 👊

**Pull request merged**